### PR TITLE
Load evolution from disk

### PR DESCRIPTION
**Pull request by caglorithm**
* Load an evolution from disk that was previously saved using `dill`.
Saving:
```python
... # evolution done here
evolution.saveEvolution("data/evolution.dill")
```
Loading, in another session:
```python
# initialize object
evolution = Evolution(lambda x: x, ParameterSpace(['mock'],  [[0, 1]]))
# load from disk
evolution = evolution.loadEvolution("data/evolution.dill")
```

Because the objects are not serializable, I needed to use `dill` for saving and loading. This means, this would be a new dependency (which I' obviously trying to avoid as much as possible). At the same time, saving and loading a previous evolution / exploration is, I think, a very important feature and reinitializing the evolution with the loaded dill is a bit non-trivial, so I thought it makes sense to solve this issue for the user instead of letting them figure it out themselves. I'd be happy about comments on this (@jajcayn?).

* Pandas DataFrame with all individuals from the entire evolution
```python
evolution.dfEvolution
```
Sorry:
Most of the changes are actually black formatting.... :( more coming very soon.

Todo:
* [x] Testing of `saveEvolution()` and `loadEvolution()`
* [ ] Lazy import of dill, is it ok? Should give `ImportError` if import fails
* [x] Docstring dfEvolution

### CODE REVIEW

File: neurolib/optimize/evolution/evolution.py [hunk 1]
```diff
@@ -700,34 +802,113 @@ def loadEvolution(self, fname):
...
+        # add fitness columns
+        # NOTE: have to do this with wvalues and divide by weights later, why?
+        # Because after loading the evolution with dill, somehow multiple fitnesses
+        # dissappear and only the first one is left. However, wvalues still has all
+        # fitnesses, and we have acces to weightList, so this hack kind of helps
+        n_fitnesses = len(self.pop[0].fitness.wvalues)
+        for i in range(n_fitnesses):
+            for ip, p in enumerate(allIndividuals):
+                column_name = "f" + str(i)
+                dfEvolution.loc[ip, column_name] = (
```
Comment (jajcayn) [inline neurolib/optimize/evolution/evolution.py:906]:
```suggestion
                dfEvolution.loc[ip,f"f{i}"] = (
```

File: neurolib/optimize/evolution/evolution.py [hunk 2]
```diff
@@ -700,34 +802,113 @@ def loadEvolution(self, fname):
...
+
+        # add fitness columns
+        # NOTE: have to do this with wvalues and divide by weights later, why?
+        # Because after loading the evolution with dill, somehow multiple fitnesses
+        # dissappear and only the first one is left. However, wvalues still has all
+        # fitnesses, and we have acces to weightList, so this hack kind of helps
+        n_fitnesses = len(self.pop[0].fitness.wvalues)
+        for i in range(n_fitnesses):
+            for ip, p in enumerate(allIndividuals):
+                column_name = "f" + str(i)
```
Comment (jajcayn) [inline neurolib/optimize/evolution/evolution.py:905]:
```suggestion
```

File: neurolib/optimize/evolution/evolutionaryUtils.py [hunk 3]
```diff
@@ -11,18 +11,32 @@
 def saveToPypet(traj, pop, gIdx):
     try:
         traj.f_add_result_group("{}.gen_{:06d}".format("evolution", gIdx))
-        traj.f_add_result("{}.gen_{:06d}.fitness".format("evolution", gIdx), np.array([p.fitness.values for p in pop]))
-        traj.f_add_result("{}.gen_{:06d}.scores".format("evolution", gIdx), np.array([p.fitness.score for p in pop]))
-        traj.f_add_result("{}.gen_{:06d}.population".format("evolution", gIdx), np.array([list(p) for p in pop]))
-        traj.f_add_result("{}.gen_{:06d}.ind_ids".format("evolution", gIdx), np.array([p.id for p in pop]))
+        traj.f_add_result(
+            "{}.gen_{:06d}.fitness".format("evolution", gIdx),
```
Comment (jajcayn) [inline neurolib/optimize/evolution/evolutionaryUtils.py:15]: I'd use either `str.format` or f-string in the whole repo, not mixing the two if not necessary...

File: neurolib/optimize/evolution/evolution.py [hunk 4]
```diff
@@ -700,34 +802,113 @@ def loadEvolution(self, fname):
         :rtype: self
         """
         import dill
+
         evolution = dill.load(open(fname, "rb"))
-        evolution.__init__(lambda x: x, self.parameterSpace)
+
+        # parameter space is not saved correctly in dill, don't know why
```
Comment (jajcayn) [inline neurolib/optimize/evolution/evolution.py:808]: this is weird...do you have any idea why? what's the problem when you load with `dill`?

File: neurolib/optimize/evolution/evolution.py [hunk 5]
```diff
@@ -678,8 +777,11 @@ def saveEvolution(self, fname=None):
         :type fname: str, optional
         """
         import dill
-        fname = fname or os.path.join("data/", "evolution-" + self.trajectoryName + ".dill")
-        dill.dump(self, open(fname, "wb"))        
+
+        fname = fname or os.path.join(
+            "data/", "evolution-" + self.trajectoryName + ".dill"
+        )
+        dill.dump(self, open(fname, "wb"))
```
Comment (jajcayn) [inline neurolib/optimize/evolution/evolution.py:784]: are you trying to dump whole class? maybe try `self.__dict__`... each class have `__dict__` method which is a dictionary of all attributes... when I am saving whole class from delfi, I save it to pickle (that shouldn't matter whether `pickle` or `dill`) with `dump(self.__dict__, fname)`

Comment (caglorithm) [reply neurolib/optimize/evolution/evolution.py:784]: pickle works only in serializable data.... I forgot which but one of the many objects (pypet?) didn't quite like to be pickled unfortunately.

File: neurolib/optimize/evolution/evolutionaryUtils.py (hunk 3 above)
Comment (caglorithm) [reply neurolib/optimize/evolution/evolutionaryUtils.py:15]: completely! this is a artifact from my pre-fstring life, oops

File: neurolib/optimize/evolution/evolution.py (hunk 5 above)
Comment (jajcayn) [reply neurolib/optimize/evolution/evolution.py:784]: yes, you're right, might be something is not pickable, then dill is indeed an option ;)
still, I'd try dilling `self.__dict__`, maybe that'll work

Comment (caglorithm) [reply neurolib/optimize/evolution/evolution.py:784]: Referring to `self.__dict__`, how would you load the original instance back into memory and replace it with `self`?

Comment (caglorithm) [reply neurolib/optimize/evolution/evolution.py:784]: Guess I'm gonna leave it as is for now, since it seems to work!

**Pull request merged**